File: kdb-wire/src/commonMain/kotlin/dev/kdb/wire/CommitPushCodec.kt
package dev.kdb.wire

import dev.kdb.document.KdbCommit

/** Length-prefixed commit list for [WireMessage.CommitPush] payloads. */
public object CommitPushCodec {
    public fun encodeCommits(commits: List<KdbCommit>): ByteArray {
        var size = 4
        val payloads = commits.map { it.toPayloadBytes() }
        for (p in payloads) {
            size += 4 + p.size
        }
        val out = ByteArray(size)
        var o = 0
        writeIntLe(out, o, commits.size)
        o += 4
        for (p in payloads) {
            writeIntLe(out, o, p.size)
            o += 4
            p.copyInto(out, o)
            o += p.size
        }
        return out
    }

    public fun decodeCommits(bytes: ByteArray): List<KdbCommit> {
        if (bytes.size < 4) return emptyList()
        var o = 0
        val count = readIntLe(bytes, o)
        o += 4
        // The count comes from a peer, so it cannot size the allocation on its own: every commit
        // costs at least its own 4-byte length prefix, so a payload of bytes.size can back at
        // most (bytes.size - 4) / 4 of them. Without this bound a four-byte CommitPush declaring
        // Int.MAX_VALUE commits asked ArrayList for a two-billion-element backing array before
        // discovering, one iteration later, that there were no commit bodies at all - an
        // OutOfMemoryError rather than a decode error, reachable by any peer that can send a
        // commitPush frame. A negative count (a length prefix with the high bit set) took out
        // ArrayList's own argument check instead. Go's DecodeCommits enforces the same bound.
        val maxPossible = (bytes.size - 4) / 4
        require(count in 0..maxPossible) {
            "commit push declares $count commits, payload can hold at most $maxPossible"
        }
        val result = ArrayList<KdbCommit>(count)
        repeat(count) {
            require(o + 4 <= bytes.size) { "truncated commit push payload" }
            val len = readIntLe(bytes, o)
            o += 4
            require(o + len <= bytes.size) { "truncated commit bytes" }
            result += KdbCommit.fromPayloadBytes(bytes.copyOfRange(o, o + len))
            o += len
        }
        return result
    }

    private fun writeIntLe(
        buf: ByteArray,
        offset: Int,
        value: Int,
    ) {
        buf[offset] = (value and 0xff).toByte()
        buf[offset + 1] = ((value shr 8) and 0xff).toByte()
        buf[offset + 2] = ((value shr 16) and 0xff).toByte()
        buf[offset + 3] = ((value shr 24) and 0xff).toByte()
    }

    private fun readIntLe(
        buf: ByteArray,
        offset: Int,
    ): Int =
        (buf[offset].toInt() and 0xff) or
            ((buf[offset + 1].toInt() and 0xff) shl 8) or
            ((buf[offset + 2].toInt() and 0xff) shl 16) or
            ((buf[offset + 3].toInt() and 0xff) shl 24)
}

File: go/kdb/wire/commit_push_codec.go
package wire

import (
	"encoding/binary"
	"fmt"

	"github.com/limidus/kdb/go/kdb/codec"
	"github.com/limidus/kdb/go/kdb/document"
)

// EncodeCommits serializes commits for CommitPush payloads.
func EncodeCommits(commits []document.Commit) ([]byte, error) {
	payloads := make([][]byte, len(commits))
	total := 4
	for i, c := range commits {
		b, err := c.ToPayloadBytes()
		if err != nil {
			return nil, err
		}
		payloads[i] = b
		total += 4 + len(b)
	}
	out := make([]byte, total)
	o := 0
	binary.LittleEndian.PutUint32(out[o:], uint32(len(commits)))
	o += 4
	for _, p := range payloads {
		binary.LittleEndian.PutUint32(out[o:], uint32(len(p)))
		o += 4
		copy(out[o:], p)
		o += len(p)
	}
	return out, nil
}

// DecodeCommits deserializes commits from CommitPush payload bytes.
func DecodeCommits(bytes []byte) ([]document.Commit, error) {
	if len(bytes) < 4 {
		return nil, nil
	}
	o := 0
	count := int(binary.LittleEndian.Uint32(bytes[o:]))
	o += 4
	// The count comes from a peer, so it cannot be used to size the allocation on its own: every
	// commit costs at least its own 4-byte length prefix, so a payload of len(bytes) can back at
	// most (len(bytes)-4)/4 of them. Without this bound a four-byte CommitPush declaring
	// 0xFFFFFFFF commits made this reserve ~800 GiB before discovering, one iteration later,
	// that there were no commit bodies at all - fatal under any memory limit (which is to say,
	// under the container and systemd deployments), and reachable by any peer that can send a
	// commitPush frame.
	if maxPossible := (len(bytes) - 4) / 4; count > maxPossible {
		return nil, fmt.Errorf("commit push declares %d commits, payload can hold at most %d", count, maxPossible)
	}
	result := make([]document.Commit, 0, count)
	for i := 0; i < count; i++ {
		if o+4 > len(bytes) {
			return nil, fmt.Errorf("truncated commit push payload")
		}
		ln := int(binary.LittleEndian.Uint32(bytes[o:]))
		o += 4
		if o+ln > len(bytes) {
			return nil, fmt.Errorf("truncated commit bytes")
		}
		c, err := document.FromPayloadBytes(bytes[o : o+ln])
		if err != nil {
			return nil, err
		}
		result = append(result, c)
		o += ln
	}
	return result, nil
}

func hashesToHex(hs []codec.Hash) []string {
	if len(hs) == 0 {
		return nil
	}
	out := make([]string, len(hs))
	for i, h := range hs {
		out[i] = h.Hex()
	}
	return out
}

func hashesFromHex(hexes []string) ([]codec.Hash, error) {
	if len(hexes) == 0 {
		return nil, nil
	}
	out := make([]codec.Hash, len(hexes))
	for i, s := range hexes {
		h, err := codec.HashFromHex(s)
		if err != nil {
			return nil, err
		}
		out[i] = h
	}
	return out, nil
}

func stubsToDto(stubs []document.CommitStub) []commitStubDto {
	if len(stubs) == 0 {
		return nil
	}
	out := make([]commitStubDto, len(stubs))
	for i, s := range stubs {
		out[i] = commitStubDto{
			OriginalHashHex: s.OriginalHash.Hex(),
			ArchiveLocation: s.ArchiveLocation,
			StubbedAtMicros: s.StubbedAt.EpochMicros(),
		}
	}
	return out
}

func stubsFromDto(dtos []commitStubDto) ([]document.CommitStub, error) {
	if len(dtos) == 0 {
		return nil, nil
	}
	out := make([]document.CommitStub, len(dtos))
	for i, d := range dtos {
		h, err := codec.HashFromHex(d.OriginalHashHex)
		if err != nil {
			return nil, err
		}
		out[i] = document.CommitStub{
			OriginalHash:    h,
			ArchiveLocation: d.ArchiveLocation,
			StubbedAt:       codec.TimestampFromEpochMicros(d.StubbedAtMicros),
		}
	}
	return out, nil
}
